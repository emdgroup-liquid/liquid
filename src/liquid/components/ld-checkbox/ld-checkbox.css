:host,
.ld-checkbox {
  /* layout */
  --ld-checkbox-size: 1.25rem;

  /* colors */
  --ld-checkbox-col-bg: var(--ld-col-wht);
  --ld-checkbox-col-bg-active: var(--ld-col-bg-g);
  --ld-checkbox-col-border: var(--ld-col-rblck4);
  --ld-checkbox-col-disabled: var(--ld-col-rblck2);
  --ld-checkbox-col-bg-disabled: var(--ld-col-rblck1);
  --ld-checkbox-dark-col-bg: var(--ld-col-bg-g);
  --ld-checkbox-dark-col-bg-hover: var(--ld-col-rblck1);
  --ld-checkbox-dark-col-bg-active: var(--ld-col-rblck2);
  --ld-checkbox-checked-col: var(--ld-col-wht);
  --ld-checkbox-col-invalid: var(--ld-col-rr-default);
  --ld-checkbox-col-invalid-hover: var(--ld-col-rr-hover);
  --ld-checkbox-col-invalid-focus: var(--ld-col-rr-focus);
  --ld-checkbox-col-invalid-active: var(--ld-col-rr-active);
  --ld-checkbox-col-warn: var(--ld-col-vy-default);
  --ld-checkbox-col-warn-hover: var(--ld-col-vy-hover);
  --ld-checkbox-col-warn-focus: var(--ld-col-vy-focus);
  --ld-checkbox-col-warn-active: var(--ld-col-vy-active);

  /* themable colors */
  --ld-checkbox-thm-col: var(--ld-col-rb-default);
  --ld-checkbox-thm-col-hover: var(--ld-col-rb-hover);
  --ld-checkbox-thm-col-active: var(--ld-col-rb-active);
  --ld-checkbox-thm-col-focus: var(--ld-col-rb-focus);

  position: relative;
  display: inline-flex;
  flex-shrink: 0;
  width: var(--ld-checkbox-size);
  height: var(--ld-checkbox-size);
  min-width: auto !important;
  box-sizing: border-box;

  input {
    appearance: none;
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    z-index: 1;
    width: 100%;
    height: 100%;
    margin: 0;

    &:checked {
      ~ .ld-checkbox__check {
        visibility: inherit;
      }
    }

    &:where(:disabled),
    &:where([aria-disabled='true']) {
      ~ .ld-checkbox__check {
        color: var(--ld-checkbox-col-disabled);
      }
      ~ .ld-checkbox__box {
        background-color: transparent;
        box-shadow: inset 0 0 0 0.1rem var(--ld-checkbox-col-bg-disabled);
      }

      &:checked {
        ~ .ld-checkbox__box {
          background-color: var(--ld-checkbox-col-bg-disabled);
        }
      }
    }

    &:where(:not(:disabled):not([aria-disabled='true'])) {
      cursor: pointer;

      ~ .ld-checkbox__check {
        color: var(--ld-checkbox-thm-col-active);
      }
      ~ .ld-checkbox__box {
        box-shadow: inset 0 0 0 0.1rem var(--ld-checkbox-thm-col);
      }

      &:focus:focus-visible {
        ~ .ld-checkbox__box {
          background-color: transparent;
          box-shadow: inset 0 0 0 0.1rem var(--ld-checkbox-thm-col-focus);
        }
      }
      &:active,
      &:active:focus-visible {
        ~ .ld-checkbox__box {
          background-color: var(--ld-checkbox-col-bg-active);
          box-shadow: inset 0 0 0 0.1rem var(--ld-checkbox-thm-col-active);
        }
      }

      &:checked {
        ~ .ld-checkbox__check {
          color: var(--ld-checkbox-checked-col);
        }
        ~ .ld-checkbox__box {
          background-color: var(--ld-checkbox-thm-col);
          box-shadow: inset 0 0 0 0.1rem var(--ld-checkbox-thm-col);
        }

        &:focus:focus-visible {
          ~ .ld-checkbox__check {
            color: var(--ld-checkbox-thm-col-active);
          }
          ~ .ld-checkbox__box {
            background-color: var(--ld-checkbox-thm-col-focus);
            box-shadow: inset 0 0 0 0.1rem var(--ld-checkbox-thm-col-focus);
          }
        }
        &:active,
        &:active:focus-visible {
          ~ .ld-checkbox__check {
            color: var(--ld-checkbox-col-bg-active);
          }
          ~ .ld-checkbox__box {
            background-color: var(--ld-checkbox-thm-col-active);
            box-shadow: inset 0 0 0 0.1rem var(--ld-checkbox-thm-col-active);
          }
        }
      }
    }
  }

  :where(input:not(:disabled):not(input[aria-disabled='true'])) {
    ~ .ld-checkbox__box {
      background-color: var(--ld-checkbox-col-bg);
    }
  }

  @media (hover: hover) {
    &:hover {
      :where(input:not(:disabled):not(input[aria-disabled='true']))
        ~ .ld-checkbox__box {
        background-color: var(--ld-checkbox-col-bg);
        box-shadow: inset 0 0 0 0.1rem var(--ld-checkbox-thm-col-hover);
      }

      :where(input:not(:disabled):not(input[aria-disabled='true'])):checked {
        ~ .ld-checkbox__check {
          color: var(--ld-checkbox-checked-col);
        }
        ~ .ld-checkbox__box {
          background-color: var(--ld-checkbox-thm-col-hover);
          box-shadow: inset 0 0 0 0.1rem var(--ld-checkbox-thm-col-hover);
        }
      }
    }
  }
}

.ld-checkbox.ld-checkbox--dark,
:host(.ld-checkbox.ld-checkbox--dark) {
  @media (hover: hover) {
    &:hover {
      :where(input:not(:disabled):not(input[aria-disabled='true']):not(:checked)) {
        ~ .ld-checkbox__box {
          background-color: var(--ld-checkbox-dark-col-bg-hover);
        }
      }
    }
  }

  input:not(:disabled):not(input[aria-disabled='true']):not(:checked) {
    ~ .ld-checkbox__box {
      background-color: var(--ld-checkbox-dark-col-bg);
    }

    &:active,
    &:active:focus-visible {
      ~ .ld-checkbox__box {
        background-color: var(--ld-checkbox-dark-col-bg-active);
      }
    }
  }
}

.ld-checkbox__check {
  visibility: hidden;
  position: absolute;
  left: 50%;
  top: 50%;
  transform: translate(-50%, -50%);
  pointer-events: none;
  z-index: 1;
}

.ld-checkbox__box {
  position: absolute;
  width: 100%;
  height: 100%;
  pointer-events: none;
  box-shadow: inset 0 0 0 0.1rem var(--ld-checkbox-col-border);
}

:host,
.ld-checkbox,
.ld-checkbox__box {
  border-radius: var(--ld-br-m);
}

:host(.ld-checkbox--highlight),
.ld-checkbox--highlight {
  --ld-checkbox-thm-col: var(--ld-checkbox-col-warn);
  --ld-checkbox-thm-col-hover: var(--ld-checkbox-col-warn-hover);
  --ld-checkbox-thm-col-active: var(--ld-checkbox-col-warn-focus);
  --ld-checkbox-thm-col-focus: var(--ld-checkbox-col-warn-active);
}

:host(.ld-checkbox--invalid),
:host(.ld-checkbox--danger),
.ld-checkbox--invalid,
.ld-checkbox--danger {
  --ld-checkbox-thm-col: var(--ld-checkbox-col-invalid);
  --ld-checkbox-thm-col-hover: var(--ld-checkbox-col-invalid-hover);
  --ld-checkbox-thm-col-active: var(--ld-checkbox-col-invalid-focus);
  --ld-checkbox-thm-col-focus: var(--ld-checkbox-col-invalid-active);
}
