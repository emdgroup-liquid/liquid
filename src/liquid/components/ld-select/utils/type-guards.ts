export const isLdOption = (
  el: HTMLElement | Node | EventTarget,
): el is HTMLLdOptionElement | HTMLLdOptionInternalElement =>
  ["LD-OPTION", "LD-OPTION-INTERNAL"].includes((el as HTMLElement)?.tagName);

export const isLdOptgroup = (
  el: HTMLElement | Node | EventTarget,
): el is HTMLLdOptgroupElement | HTMLLdOptgroupInternalElement =>
  ["LD-OPTGROUP", "LD-OPTGROUP-INTERNAL"].includes(
    (el as HTMLElement)?.tagName,
  );

export const isLdOptionInternal = (
  el: HTMLElement | Node | EventTarget,
): el is HTMLLdOptionInternalElement =>
  ["LD-OPTION-INTERNAL"].includes((el as HTMLElement)?.tagName);

export const isLdOptgroupInternal = (
  el: HTMLElement | Node | EventTarget,
): el is HTMLLdOptgroupInternalElement =>
  ["LD-OPTGROUP-INTERNAL"].includes((el as HTMLElement)?.tagName);

type HTMLLdOptInternal =
  | HTMLLdOptionInternalElement
  | HTMLLdOptgroupInternalElement;
export const isLdOptInternalHidden = (
  opt: HTMLLdOptInternal,
): opt is
  | (HTMLLdOptInternal & {
      hidden: true;
    })
  | (HTMLLdOptInternal & {
      filtered: true;
    }) => {
  return opt.hidden || opt.filtered;
};
