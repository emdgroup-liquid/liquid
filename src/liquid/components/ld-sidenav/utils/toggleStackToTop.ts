import { LdSidenavNavitem } from '../ld-sidenav-navitem/ld-sidenav-navitem'

export const toggleStackToTop = (el: HTMLElement, stacked: boolean) => {
  const outerSlider = el.closest('ld-sidenav-slider')
  if (!outerSlider) return

  const sidenavPaddingY =
    parseInt(
      window
        .getComputedStyle(outerSlider)
        .getPropertyValue('--ld-sidenav-padding-y')
    ) * 16

  // If not stacked, put everything back in place.
  if (!stacked) {
    el.style.removeProperty('--ld-slider-navitem-move-up')
    return
  }

  // Else, calculate how much vertical space is taken by stacked items above
  // and move the nav item up accordingly.
  let totalSpaceAbove = 0
  let totalSpaceOccupiedAbove = 0
  for (const elem of Array.from(el.parentElement.children)) {
    if (el === elem) break

    if (
      elem.tagName === 'LD-SIDENAV-SEPARATOR' ||
      (elem.tagName === 'LD-SIDENAV-NAVITEM' &&
        !['secondary', 'tertiary'].includes(
          (elem as unknown as LdSidenavNavitem).mode
        ))
    ) {
      totalSpaceOccupiedAbove += elem.getBoundingClientRect().height
    } else {
      totalSpaceAbove += sidenavPaddingY
    }
    totalSpaceAbove += elem.getBoundingClientRect().height
  }

  const spaceToMoveUp = totalSpaceAbove - totalSpaceOccupiedAbove
  el.style.setProperty('--ld-slider-navitem-move-up', `-${spaceToMoveUp}px`)
}
