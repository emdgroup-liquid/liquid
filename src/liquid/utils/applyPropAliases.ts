/**
 * Allows using lowercase props as aliases for camelcase props.
 * This resolves issues in react based applications, where
 * otherwise only kebabcase attributes would work, but are not
 * proposed by intellisense.
 */
export function applyPropAliases() {
  if (!this.el) {
    throw new Error(`Component ${this.name} is missing @Element() decorator.`)
  }
  for (const prop in this) {
    if (prop !== prop.toLowerCase()) {
      const lowercaseAttr = this.el.getAttribute(prop.toLowerCase())
      if (lowercaseAttr) {
        this[prop] = lowercaseAttr
      }
    }
  }
}
