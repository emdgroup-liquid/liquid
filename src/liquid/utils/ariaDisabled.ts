// 'true', '1', 'yolo' etc. correspond to an aria disabled element, but not 'false'.
export const isAriaDisabled = (ariaDisabledValue: string | undefined) =>
  ariaDisabledValue && ariaDisabledValue !== "false";
