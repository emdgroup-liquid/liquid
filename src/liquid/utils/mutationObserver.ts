import MutationObserver from "mutation-observer";

let triggerableMutationObservers = [];
const TriggerableMutationObserver = function (cb) {
  const triggerableMutationObserver = new MutationObserver(cb);
  triggerableMutationObserver.trigger = cb;
  triggerableMutationObservers.push(triggerableMutationObserver);
  return triggerableMutationObserver;
};

global.MutationObserver = TriggerableMutationObserver as MutationObserver;

export function getTriggerableMutationObservers() {
  return triggerableMutationObservers;
}

export function clearTriggerableMutationObservers() {
  triggerableMutationObservers = [];
}
