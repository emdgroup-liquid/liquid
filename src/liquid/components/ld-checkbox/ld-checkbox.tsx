import {
  Component,
  Element,
  Event,
  EventEmitter,
  h,
  Host,
  Method,
  Prop,
  State,
  Watch,
} from '@stencil/core'
import { cloneAttributes } from '../../utils/cloneAttributes'
import { getClassNames } from '../../utils/getClassNames'
import { registerAutofocus } from '../../utils/focus'

/**
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 * @part input - Actual input element
 */
@Component({
  tag: 'ld-checkbox',
  styleUrl: 'ld-checkbox.css',
  shadow: true,
})
export class LdCheckbox implements InnerFocusable, ClonesAttributes {
  @Element() el: HTMLInputElement

  private attributesObserver: MutationObserver

  private input: HTMLInputElement
  private hiddenInput: HTMLInputElement

  /** Automatically focus the form control when the page is loaded. */
  @Prop({ reflect: true }) autofocus: boolean

  /** Indicates whether the checkbox is checked. */
  @Prop({ mutable: true }) checked? = false

  /** Disabled state of the checkbox. */
  @Prop() disabled?: boolean

  /** Associates the control with a form element. */
  @Prop() form?: string

  /**
   * Set this property to `true` to indicate that the checkbox's value is neither true nor false.
   * The prop is removed automatically as soon as the checkbox is clicked (if not disabled).
   */
  @Prop({ mutable: true }) indeterminate?: boolean

  /** Set this property to `true` in order to mark the checkbox visually as invalid. */
  @Prop() invalid?: boolean

  /** Tab index of the input. */
  @Prop() ldTabindex?: number

  /** Display mode. */
  @Prop() mode?: 'highlight' | 'danger'

  /** Used to specify the name of the control. */
  @Prop() name?: string

  /** The value is not editable. */
  @Prop() readonly?: boolean

  /** Set this property to `true` in order to mark the checkbox as required. */
  @Prop() required?: boolean

  /** Checkbox tone. Use `'dark'` on white backgrounds. Default is a light tone. */
  @Prop() tone?: 'dark'

  /** The input value. */
  @Prop() value?: string

  @State() clonedAttributes

  /** Emitted when the input value changed and the element loses focus. */
  @Event() ldchange: EventEmitter<boolean>

  /** Emitted when the input value changed. */
  @Event() ldinput: EventEmitter<boolean>

  /** Sets focus on the checkbox. */
  @Method()
  async focusInner() {
    if (this.input !== undefined) {
      this.input.focus()
    }
  }

  @Watch('checked')
  updateIndeterminate() {
    this.indeterminate = undefined
  }

  @Watch('checked')
  @Watch('form')
  @Watch('name')
  @Watch('value')
  updateHiddenInput() {
    const outerForm = this.el.closest('form')
    if (!this.hiddenInput && this.name && (outerForm || this.form)) {
      this.createHiddenInput()
    }

    if (this.hiddenInput) {
      if (!this.name) {
        this.hiddenInput.remove()
        this.hiddenInput = undefined
        return
      }

      this.hiddenInput.name = this.name
      this.hiddenInput.checked = this.checked

      if (this.value) {
        this.hiddenInput.value = this.value
      } else {
        this.hiddenInput.removeAttribute('value')
      }

      if (this.form) {
        this.hiddenInput.setAttribute('form', this.form)
      } else if (this.hiddenInput.getAttribute('form')) {
        if (outerForm) {
          this.hiddenInput.removeAttribute('form')
        } else {
          this.hiddenInput.remove()
          this.hiddenInput = undefined
        }
      }
    }
  }

  private createHiddenInput() {
    this.hiddenInput = document.createElement('input')
    this.hiddenInput.type = 'checkbox'
    this.hiddenInput.style.visibility = 'hidden'
    this.hiddenInput.style.position = 'absolute'
    this.hiddenInput.style.pointerEvents = 'none'
    this.el.appendChild(this.hiddenInput)
  }

  private handleChange = (ev: InputEvent) => {
    this.el.dispatchEvent(new InputEvent('change', ev))
    this.ldchange.emit(this.checked)
  }

  private handleClick = (ev: MouseEvent) => {
    if (this.disabled || this.el.getAttribute('aria-disabled') === 'true') {
      ev.preventDefault()
      return
    }

    this.checked = !this.checked

    if (!ev.isTrusted) {
      // This happens, when a click event is dispatched on the host element
      // from the outside i.e. on click on a parent ld-label element.
      this.el.dispatchEvent(
        new InputEvent('input', { bubbles: true, composed: true })
      )
      this.handleInput()
      this.el.dispatchEvent(new InputEvent('change', { bubbles: true }))
      this.ldchange.emit(this.checked)
    }
  }

  private handleInput = () => {
    this.ldinput.emit(this.checked)
  }

  componentWillLoad() {
    this.attributesObserver = cloneAttributes.call(this, ['tone', 'mode'])

    const outerForm = this.el.closest('form')

    if (this.name && (outerForm || this.form)) {
      this.createHiddenInput()
      this.hiddenInput.checked = this.checked
      this.hiddenInput.name = this.name

      if (this.form) {
        this.hiddenInput.setAttribute('form', this.form)
      }

      if (this.value) {
        this.hiddenInput.value = this.value
      }
    }

    registerAutofocus(this.autofocus)
  }

  disconnectedCallback() {
    /* istanbul ignore if */
    if (this.attributesObserver) this.attributesObserver.disconnect()
  }

  render() {
    const cl = [
      'ld-checkbox',
      this.mode && `ld-checkbox--${this.mode}`,
      this.tone && `ld-checkbox--${this.tone}`,
      this.invalid && 'ld-checkbox--invalid',
    ]

    return (
      <Host part="root" class={getClassNames(cl)} onClick={this.handleClick}>
        <input
          type="checkbox"
          {...this.clonedAttributes}
          checked={this.checked}
          disabled={this.disabled}
          indeterminate={this.indeterminate}
          onChange={this.handleChange}
          onInput={this.handleInput}
          part="input focusable"
          ref={(ref) => (this.input = ref)}
          tabIndex={this.ldTabindex}
          value={this.value}
        />
        {/* custom icon check */}
        <svg
          class="ld-checkbox__check"
          part="check"
          width="14"
          height="14"
          fill="none"
          viewBox="0 0 14 14"
        >
          <path
            d="M12 4L5.40795 10L2 6.63964"
            stroke="currentColor"
            stroke-width="3"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
        <div class="ld-checkbox__box" part="box"></div>
      </Host>
    )
  }
}
