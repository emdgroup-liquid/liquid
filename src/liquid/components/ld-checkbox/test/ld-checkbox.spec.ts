import { newSpecPage } from '@stencil/core/testing'
import { LdCheckbox } from '../ld-checkbox'
import '../../../utils/mutationObserver'

describe('ld-checkbox', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [LdCheckbox],
      html: `<ld-checkbox></ld-checkbox>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('disabled', async () => {
    const page = await newSpecPage({
      components: [LdCheckbox],
      html: `<ld-checkbox disabled></ld-checkbox>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('heighlight', async () => {
    const page = await newSpecPage({
      components: [LdCheckbox],
      html: `<ld-checkbox mode="highlight"></ld-checkbox>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('danger', async () => {
    const page = await newSpecPage({
      components: [LdCheckbox],
      html: `<ld-checkbox mode="danger"></ld-checkbox>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('tone', async () => {
    const page = await newSpecPage({
      components: [LdCheckbox],
      html: `<ld-checkbox tone="dark"></ld-checkbox>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('updates checked prop on checked value change', async () => {
    const page = await newSpecPage({
      components: [LdCheckbox],
      html: `<ld-checkbox></ld-checkbox>`,
    })
    const ldCheckbox = page.root
    expect(ldCheckbox.checked).toBe(false)

    const input = ldCheckbox.shadowRoot.querySelector('input')
    expect(input.checked).toBe(false)

    ldCheckbox.click()
    await page.waitForChanges()

    expect(ldCheckbox.checked).toBe(true)
    expect(input.checked).toBe(true)
  })

  it('prevents changes if disabled', async () => {
    const page = await newSpecPage({
      components: [LdCheckbox],
      html: `<ld-checkbox disabled></ld-checkbox>`,
    })
    const ldCheckbox = page.root
    expect(ldCheckbox.checked).toBe(false)

    const input = ldCheckbox.shadowRoot.querySelector('input')
    expect(input.checked).toBe(false)

    ldCheckbox.click()
    await page.waitForChanges()

    expect(ldCheckbox.checked).toBe(false)
    expect(input.checked).toBe(false)
  })

  it('prevents changes if aria-disabled', async () => {
    const page = await newSpecPage({
      components: [LdCheckbox],
      html: `<ld-checkbox aria-disabled="true"></ld-checkbox>`,
    })
    const ldCheckbox = page.root
    expect(ldCheckbox.checked).toBe(false)

    const input = ldCheckbox.shadowRoot.querySelector('input')
    expect(input.checked).toBe(false)

    ldCheckbox.click()
    await page.waitForChanges()

    expect(ldCheckbox.checked).toBe(false)
    expect(input.checked).toBe(false)
  })

  // TODO: Uncomment, as soon as Stencil's JSDom implementation
  // supports bubbling of composed events into the light DOM.
  xit('emits focus and blur events', async () => {
    const page = await newSpecPage({
      components: [LdCheckbox],
      html: `<ld-checkbox></ld-checkbox>`,
    })
    const ldCheckbox = page.root
    const input = ldCheckbox.shadowRoot.querySelector('input')

    const focusHandler = jest.fn()
    ldCheckbox.addEventListener('focus', focusHandler)
    input.dispatchEvent(new FocusEvent('focus'))
    jest.advanceTimersByTime(0)
    expect(focusHandler).toHaveBeenCalled()

    const blurHandler = jest.fn()
    ldCheckbox.addEventListener('blur', blurHandler)
    input.dispatchEvent(new FocusEvent('blur'))
    jest.advanceTimersByTime(0)
    expect(blurHandler).toHaveBeenCalled()
  })

  it('emits change and ldchange events', async () => {
    const page = await newSpecPage({
      components: [LdCheckbox],
      html: `<ld-checkbox></ld-checkbox>`,
    })
    const ldCheckbox = page.root
    const input = ldCheckbox.shadowRoot.querySelector('input')

    const changeHandler = jest.fn()
    ldCheckbox.addEventListener('change', changeHandler)
    const ldchangeHandler = jest.fn()
    ldCheckbox.addEventListener('ldchange', ldchangeHandler)

    input.dispatchEvent(new InputEvent('change', { bubbles: true }))
    expect(changeHandler).toHaveBeenCalled()
    expect(ldchangeHandler).toHaveBeenCalled()
  })

  // TODO: Uncomment, as soon as Stencil's JSDom implementation
  // supports bubbling of composed events into the light DOM.
  xit('emits input event', async () => {
    const page = await newSpecPage({
      components: [LdCheckbox],
      html: `<ld-checkbox></ld-checkbox>`,
    })
    const ldCheckbox = page.root
    const input = ldCheckbox.shadowRoot.querySelector('input')

    const inputHandler = jest.fn()
    ldCheckbox.addEventListener('input', inputHandler)

    input.dispatchEvent(
      new InputEvent('input', { bubbles: true, composed: true })
    )
    expect(inputHandler).toHaveBeenCalled()
  })

  it('emits ldinput event', async () => {
    const page = await newSpecPage({
      components: [LdCheckbox],
      html: `<ld-checkbox></ld-checkbox>`,
    })
    const ldCheckbox = page.root
    const input = ldCheckbox.shadowRoot.querySelector('input')

    const ldinputHandler = jest.fn()
    ldCheckbox.addEventListener('ldinput', ldinputHandler)

    input.dispatchEvent(
      new InputEvent('input', { bubbles: true, composed: true })
    )
    expect(ldinputHandler).toHaveBeenCalled()
  })

  it('emits events on programmatic click', async () => {
    const page = await newSpecPage({
      components: [LdCheckbox],
      html: `<ld-checkbox></ld-checkbox>`,
    })
    const ldCheckbox = page.root
    const changeHandler = jest.fn()
    const ldchangeHandler = jest.fn()
    const inputHandler = jest.fn()
    const ldinputHandler = jest.fn()

    ldCheckbox.addEventListener('change', changeHandler)
    ldCheckbox.addEventListener('ldchange', ldchangeHandler)
    ldCheckbox.addEventListener('input', inputHandler)
    ldCheckbox.addEventListener('ldinput', ldinputHandler)
    ldCheckbox.click()

    expect(changeHandler).toHaveBeenCalled()
    expect(ldchangeHandler).toHaveBeenCalled()
    expect(inputHandler).toHaveBeenCalled()
    expect(ldinputHandler).toHaveBeenCalled()
  })

  it('allows to set inner focus', async () => {
    const page = await newSpecPage({
      components: [LdCheckbox],
      html: `<ld-checkbox />`,
    })
    const ldCheckbox = page.root
    const input = ldCheckbox.shadowRoot.querySelector('input')

    input.focus = jest.fn()
    await ldCheckbox.focusInner()

    expect(input.focus).toHaveBeenCalled()
  })

  it('creates hidden input field, if inside a form', async () => {
    const { root } = await newSpecPage({
      components: [LdCheckbox],
      html: `<form><ld-checkbox name="example" /></form>`,
    })
    expect(root).toMatchSnapshot()
  })

  it('sets initial state on hidden input', async () => {
    const page = await newSpecPage({
      components: [LdCheckbox],
      html: '<form><ld-checkbox name="example" checked value="yolo" /></form>',
    })
    const ldCheckbox = page.root
    expect(ldCheckbox.querySelector('input')).toHaveProperty('name', 'example')
    expect(ldCheckbox.querySelector('input')).toHaveProperty('value', 'yolo')
  })

  it('updates hidden input field', async () => {
    const { root, waitForChanges } = await newSpecPage({
      components: [LdCheckbox],
      html: `<form><ld-checkbox name="example" /></form>`,
    })
    const ldCheckbox = root
    await waitForChanges()

    expect(ldCheckbox.querySelector('input')).toHaveProperty('name', 'example')
    expect(ldCheckbox.querySelector('input')).toHaveProperty('checked', false)

    ldCheckbox.setAttribute('name', 'test')
    await waitForChanges()

    expect(ldCheckbox.querySelector('input')).toHaveProperty('name', 'test')

    ldCheckbox.removeAttribute('name')
    await waitForChanges()

    expect(ldCheckbox.querySelector('input')).toEqual(null)

    ldCheckbox.setAttribute('name', 'test')
    await waitForChanges()

    expect(ldCheckbox.querySelector('input')).toHaveProperty('name', 'test')

    ldCheckbox.dispatchEvent(new Event('click'))
    await waitForChanges()

    expect(ldCheckbox.querySelector('input')).toHaveProperty('checked', true)

    ldCheckbox.setAttribute('value', 'test')
    await waitForChanges()

    expect(ldCheckbox.querySelector('input')).toHaveProperty('value', 'test')

    ldCheckbox.setAttribute('form', 'my-form')
    await waitForChanges()

    expect(ldCheckbox.querySelector('input')).toHaveProperty('form', 'my-form')

    ldCheckbox.removeAttribute('form')
    await waitForChanges()

    expect(ldCheckbox.querySelector('input').getAttribute('form')).toEqual(null)

    ldCheckbox.removeAttribute('value')
    ldCheckbox.value = undefined
    await waitForChanges()

    expect(ldCheckbox.querySelector('input').getAttribute('value')).toEqual(
      null
    )
  })

  it('removes hidden input field with removal of form prop when there is no outer form', async () => {
    const { root, waitForChanges } = await newSpecPage({
      components: [LdCheckbox],
      html: `<ld-checkbox name="example" form="my-form" />`,
    })
    const ldCheckbox = root
    await waitForChanges()

    expect(ldCheckbox.querySelector('input')).toHaveProperty('form', 'my-form')

    ldCheckbox.removeAttribute('form')
    await waitForChanges()

    expect(ldCheckbox.querySelectorAll('input').length).toEqual(0)
  })

  it('uses hidden input field with referenced form', async () => {
    const { root, waitForChanges } = await newSpecPage({
      components: [LdCheckbox],
      html: '<ld-checkbox name="example" form="yolo" />',
    })
    const ldCheckbox = root
    await waitForChanges()

    expect(ldCheckbox.querySelector('input')).toHaveProperty('name', 'example')
  })

  describe('autofocus', () => {
    afterEach(() => {
      jest.restoreAllMocks()
    })

    it('sets focus if element is first with autofocus in DOM', async () => {
      const page = await newSpecPage({
        components: [LdCheckbox],
        html: `<ld-checkbox autofocus />`,
      })
      const ldCheckbox = page.root as HTMLLdCheckboxElement
      const input = ldCheckbox.shadowRoot.querySelector('input')
      jest
        .spyOn(document, 'querySelectorAll')
        .mockImplementation(() => page.body.querySelectorAll('*'))

      input.focus = jest.fn()

      jest.advanceTimersByTime(200)
      await page.waitForChanges()

      expect(input.focus).toHaveBeenCalled()
    })

    it('does not set focus if element is not first with autofocus in DOM', async () => {
      const page = await newSpecPage({
        components: [LdCheckbox],
        html: `
          <form>
            <input type="text" name="yolo" autofocus />
            <ld-checkbox autofocus />
          </form>`,
      })
      const ldCheckbox = page.root as HTMLLdCheckboxElement
      const input = ldCheckbox.shadowRoot.querySelector('input')
      jest
        .spyOn(document, 'querySelectorAll')
        .mockImplementation(() => page.body.querySelectorAll('form *'))

      input.focus = jest.fn()

      jest.advanceTimersByTime(200)
      await page.waitForChanges()

      expect(input.focus).not.toHaveBeenCalled()
    })
  })
})
