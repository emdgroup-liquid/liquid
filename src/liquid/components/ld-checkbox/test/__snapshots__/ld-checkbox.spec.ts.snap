// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-checkbox creates hidden input field, if inside a form 1`] = `
<ld-checkbox class="ld-checkbox" part="root">
  <mock:shadow-root>
    <input part="input focusable" type="checkbox">
    <svg class="ld-checkbox__check" fill="none" height="14" part="check" viewBox="0 0 14 14" width="14" xmlns="http://www.w3.org/2000/svg">
      <path d="M12 4L5.40795 10L2 6.63964" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="3"></path>
    </svg>
    <div class="ld-checkbox__box" part="box"></div>
  </mock:shadow-root>
  <input type="hidden">
</ld-checkbox>
`;

exports[`ld-checkbox danger 1`] = `
<ld-checkbox class="ld-checkbox ld-checkbox--danger" mode="danger" part="root">
  <mock:shadow-root>
    <input part="input focusable" type="checkbox">
    <svg class="ld-checkbox__check" fill="none" height="14" part="check" viewBox="0 0 14 14" width="14" xmlns="http://www.w3.org/2000/svg">
      <path d="M12 4L5.40795 10L2 6.63964" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="3"></path>
    </svg>
    <div class="ld-checkbox__box" part="box"></div>
  </mock:shadow-root>
</ld-checkbox>
`;

exports[`ld-checkbox disabled 1`] = `
<ld-checkbox class="ld-checkbox" disabled="" part="root">
  <mock:shadow-root>
    <input disabled="" part="input focusable" type="checkbox">
    <svg class="ld-checkbox__check" fill="none" height="14" part="check" viewBox="0 0 14 14" width="14" xmlns="http://www.w3.org/2000/svg">
      <path d="M12 4L5.40795 10L2 6.63964" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="3"></path>
    </svg>
    <div class="ld-checkbox__box" part="box"></div>
  </mock:shadow-root>
</ld-checkbox>
`;

exports[`ld-checkbox heighlight 1`] = `
<ld-checkbox class="ld-checkbox ld-checkbox--highlight" mode="highlight" part="root">
  <mock:shadow-root>
    <input part="input focusable" type="checkbox">
    <svg class="ld-checkbox__check" fill="none" height="14" part="check" viewBox="0 0 14 14" width="14" xmlns="http://www.w3.org/2000/svg">
      <path d="M12 4L5.40795 10L2 6.63964" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="3"></path>
    </svg>
    <div class="ld-checkbox__box" part="box"></div>
  </mock:shadow-root>
</ld-checkbox>
`;

exports[`ld-checkbox renders 1`] = `
<ld-checkbox class="ld-checkbox" part="root">
  <mock:shadow-root>
    <input part="input focusable" type="checkbox">
    <svg class="ld-checkbox__check" fill="none" height="14" part="check" viewBox="0 0 14 14" width="14" xmlns="http://www.w3.org/2000/svg">
      <path d="M12 4L5.40795 10L2 6.63964" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="3"></path>
    </svg>
    <div class="ld-checkbox__box" part="box"></div>
  </mock:shadow-root>
</ld-checkbox>
`;

exports[`ld-checkbox tone 1`] = `
<ld-checkbox class="ld-checkbox ld-checkbox--dark" part="root" tone="dark">
  <mock:shadow-root>
    <input part="input focusable" type="checkbox">
    <svg class="ld-checkbox__check" fill="none" height="14" part="check" viewBox="0 0 14 14" width="14" xmlns="http://www.w3.org/2000/svg">
      <path d="M12 4L5.40795 10L2 6.63964" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="3"></path>
    </svg>
    <div class="ld-checkbox__box" part="box"></div>
  </mock:shadow-root>
</ld-checkbox>
`;
